hazelcast/src/hazelcast/client/network.cpp

```diff
@@ -767,6 +740,18 @@ namespace hazelcast {
                 return on_authenticated(connection, result);
             }
 
+            address ClientConnectionManagerImpl::translate(const member &m) {
+                if (use_public_address_) {
+                    auto public_addr_it = m.address_map().find(PUBLIC_ENDPOINT_QUALIFIER);
+                    if (public_addr_it != m.address_map().end()) {
+                        return public_addr_it->second;
+                    }
+                    return m.get_address();
+                }
+
+                return *address_provider_->translate(m.get_address());
```

Comment:
They are exclusive. The Java side has [this line](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/impl/spi/impl/TranslateToPublicAddressProvider.java#L66) for the check. I added a method `bool is_default_provider()` to the `AddressProvider` interface so that we can do this check while constructing the connection manager and i set the `use_public_address_` bool flag according to it to be consistent with Java.